File: LC102.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        q = []
        res = []

        if not root:
            return res
        
        q.append(root)
        q.append(None)

        while q[0] != None:

            t = q[0]
            v = []
            while t:
                q.pop(0)
                v.append(t.val)

                if t.left:
                    q.append(t.left)
                if t.right:
                    q.append(t.right)
                
                t = q[0]
            
            res.append(v)
            q.pop(0)
            q.append(None)
        
        return res
                

File: LC102.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {

        vector<vector<int>> res;

        if(!root)
            return res;
        
        queue <TreeNode*> q;

        q.push(root);
        q.push(NULL);
        
        while(q.front()){
            
            vector <int> v;
            TreeNode* t = q.front();
            
            while(t){
                q.pop();
                v.push_back(t->val);
            
                if(t->left)
                    q.push(t->left);
                if(t->right)
                    q.push(t->right);
                t = q.front();
            }
            
            res.push_back(v);
            
            q.pop();
            
            q.push(NULL);
        }

        return res;
    }
};
